golang/src/memory.rs

```diff
@@ -0,0 +1,69 @@
+use std::mem;
+use std::slice;
+
+#[derive(Copy, Clone)]
+#[repr(C)]
+pub struct Buffer {
+    pub ptr: *mut u8,
+    pub len: usize,
+    pub cap: usize,
+}
+
+// this frees memory we released earlier
+#[no_mangle]
+pub extern "C" fn free_buffer(buf: Buffer) {
+    unsafe {
+        let _ = buf.consume();
+    }
+}
+
+impl Buffer {
+    // read provides a reference to the included data to be parsed or copied elsewhere
+    // data is only guaranteed to live as long as the Buffer
+    // (or the scope of the extern "C" call it came from)
```

Comment:
This does not describe a safe function. I think this will have to be marked as unsafe. 
We may want to think about a way to reduce the unsafe surface. IE maybe doing some copying up front so that the bulk of the code can be safe.